### ISSUE TITLE
[Rebalancing] Smarter allocation of ECS resources

### ISSUE THREAD

**Issue by abby-fuller** [aws/containers-roadmap#105]
@euank 

Doesn't seem like ECS rebalances tasks to allocate resources more effectively.  For example,  if I have task A and task B running on different cluster hosts, and try to deploy task C, I'll get a resource error, and task C will fail to deploy.  However, if ECS rebalanced tasks A and B to run on the same box, there would be enough resources to deploy task C.

This comes up pretty often for us-  with the current placement of tasks, we run of resources and cannot deploy/achieve 100% utilization of our cluster hosts, because things are balanced inefficiently.  Max, we're probably getting 75% utilization, and 25% is going to waste.

**Comment by vimmis** [comment]
+1.
I still run out of enough CPU or Memory to place a task, even though the whole cluster is still having 2 or 3 instances worth CPU and Memory free. Currently, I have to scale down to rebalance tasks manually and then scale up to allow recycling of big tasks.

**Comment by whereisaaron** [comment]
If there isn't a rebalancing tool for ECS right now you could consider other AWS containers options:

**Fargate**, then node management/packing isn't your problem (for a price admittedly) 😄 

**EKS** where you have a lot more scheduling tools, e.g. you can deploy the [`descheduler`](https://github.com/kubernetes-incubator/descheduler) to repack/reschedule workloads to avoid under/over utlitised nodes, and the [`cluster-autoscaler`](https://github.com/kubernetes/autoscaler/tree/master/cluster-autoscaler) which will manage the ASG size in a workload-aware manner, picking the best nodes to remove for scale-in, and reschedule work-loads off low utilization nodes to enable scale-in, or pre-scale-up to ensure reserve capacity for fast service scale-up. You can combine those tools with [Pod Disruption Budgets](https://kubernetes.io/docs/tasks/run-application/configure-pdb/) to ensure replicas don't get rescheduled too frequently or too quickly. And both [soft and hard CPU/RAM resource requirements](https://kubernetes.io/docs/tasks/configure-pod-container/assign-memory-resource/) at both the replica and application (namespace) level, enabling strict reservation or sharing CPU/RAM between burstable workloads. You generally just have a lot more workload management tools and options with EKS.

For ECS, perhaps you need a service like the Kubernetes `descheduler` to facilitate repacking?

**Comment by vimmis** [comment]
> For ECS, perhaps you need a service like the Kubernetes descheduler to facilitate repacking?
@whereisaaron, yes at least for my use case.

**Comment by benyanke** [comment]
Any updates on this front?

**Comment by mdc-drivin** [comment]
+1 for rebalancing.  We currently us AZ spread for placement, but we have automated autoscaling with with rolling our own ami to keep our hosts as up to date as possible. When we drain nodes from the cluster, tasks have to be moved to other nodes, and when autoscaling brings up new nodes, those nodes will sit idle until either another node is removed or all the services have been redployed over time.

**Comment by asdf01** [comment]
WTF (Why The Face) AWS!  It's coming up to 4 years.  4 ducking years since this issue was first raised.  The situation today is even worse than 4 years ago.  

Not everyone desperately waiting on this issue is actively commenting and begging in the comments.  A lot of people are probably like me, waiting patiently on the sidelines hoping AWS would eventually come to their senses and fix this issue.  

We used to be able to work around this issue ourselves by starting up a new set of container hosts.  (Which you guys confusingly named "container instances").  By gradually draining and deregistering the old set of 3x container hosts, our 400x or so docker containers would gradually get distributed evenly over the new set of 3x container hosts.  But somehow, instead of fixing this issue, you guys have recently managed to break this workaround.

Just look at this screenshot:
![image](https://user-images.githubusercontent.com/19940508/58223840-70462d80-7d6f-11e9-9c70-3a2955be6aac.png)
Look at it!  This is the type of sheet I have to deal with now.  Through deregistering a container host that was previously struggling to run 190x or so containers, out of the 3x new container hosts, one got allocated 40x or so tasks, one got allocated 60x or so tasks, the last one got allocated 0.  Ducking zero!  

The only reason I'm changing out the container hosts for new ones is because the 3x old container hosts had unequal distribution and errors were occurring.  One was struggling with memory and started to page at 190x or so containers.  While one was relatively idle with only 110x or so containers.  With all the traffic load balanced over containers in all 3x containers hosts, the net effect is errors.  

Add the ducking rebalancing / redistribution feature.  You're making us all look like idiots in front of our employers.  We fought for AWS to be adopted over every other cloud offering.  AWS used to be the innovators.  What happened? 

Microsoft Azure is fighting hard for our business from the top down.  I used to have a lot of technological reasons to fight for AWS against Azure.  With issues such as this unfixed for 4 ducking years, I'm going to look like a clown if I continue arguing to stay with AWS.  

Wake up.

**Comment by asdf01** [comment]
What a waste of time:
![image](https://user-images.githubusercontent.com/19940508/58225893-42b1b200-7d78-11e9-8f11-a267c9598576.png)
One container host got allocated 176 containers, while another got allocated with less than half at 83.  AZ spread my butter.

**Comment by Zebirdman** [comment]
This is a massive deal and im really surprised to see that ECS has no intelligent balancing with container placement on instances. 

+1 for this it would increase the value of ECS hugely.

**Comment by Zebirdman** [comment]
Update for anyone who lands here, i used the lambda function shown here https://github.com/aws-samples/ecs-refarch-task-rebalancing and was able to get some pretty reasonable rebalancing across our instances.

You can also modify the code so that instead of it creating new task definitions it uses the force-update-true setting to simply re pull the existing versions. For now ive found that this is enough for our needs but hopefully in future this will be a built in feature without needing external monitors and lambda functions to trigger.

**Comment by dimisjim** [comment]
@Zebirdman according to https://github.com/aws-samples/ecs-refarch-task-rebalancing/issues/2 it works only with up to 10 services, making it a non-solution really : /

**Comment by cmeisinger** [comment]
I'm in the middle of an ECS rollout myself and just ran into this.   Are there any plans to implement appropriate scaling policies for ASG based ECS deployments?

**Comment by coultn** [comment]
> 
> @coultn
> 
> Issues/limitations I have encountered so far:
> 
>     1. Agents disconnecting due to OOM. This could happen. Problem is that the cluster has almost no way to tackle this and recover. As long as the agent is permanently disconnected, the scheduler doesn't kill it or try to help it recover in any way. This renders all tasks already running there invisible to the scheduler. There might be still running, there might be pending tasks, but there is no way for the scheduler to know.
> 
>     2. Scaling out with a scaling policy in the auto scaling group. Sure, it will spawn a new node, but that new node won't receive any existing tasks that already run in some other instance, thus rendering it kind of pointless.
> 
>     3. Assuming a node running hot at close to 100% CPU usage, it could be really helpful to have a mechanism to detect which task has an abnormally high CPUUtilization, in order to stop it and place it in another node that could accommodate its recent high utilization.
> 
> 
> Any plan currently to address the above issues?

Thanks for your detailed feedback!  These really are three different issues.  The first one has to do with agent health generally, but not specifically rebalancing.  We are always working to make the agent more reliable and robust.

Regarding the second issue:  this has to do with the fact that cluster autoscaling today is usually reactive, meaning it only scales in response to containers that are already running.  We are planning on some improvements to cluster autoscaling (https://github.com/aws/containers-roadmap/issues/76) that will mitigate this, by scaling the cluster based on the containers you want to run, not just the containers that are already running.  This means that you will be more likely to place new tasks on newly-added instances.  Additionally, as the status of this issue indicates, we are researching how to rebalance services to that have spread placement strategies defined so that the service can be spread evenly in the manner intended even if it is not currently evenly spread.

The third issue is what I would call "cluster heat management."  This would be a new capability for ECS that goes beyond just placement strategies.   The problem is fairly complicated, because there may well be multiple tasks and multiple instances with high utilization, and those utilizations may be changing rapidly.  Optimal solutions to this type of problem are computationally infeasible (they boil down to problems know as "binpacking" or "knapsack" problems, which are known to be computationally infeasible to solve optimally).   Thus, solutions to such problems usually rely on approximations and heuristics.

**Comment by Zebirdman** [comment]
@dimisjim i wasn't aware of that but the lambda function is using the generic describe services API call. It would be trivial to change this to individual service calls based on a list you pass to it which should easily get around this. Of course im not aware of how many services people might be running, based on what i do it doesn't seem like a serious inconvienience to do this.

**Comment by fpuntoni** [comment]
+1 rebalancing

**Comment by asdf01** [comment]
♫Welcome to the jungle, it gets worse here every day♫
![image](https://user-images.githubusercontent.com/19940508/62349590-50439f00-b554-11e9-84cb-95c74e649526.png)

**Comment by coultn** [comment]
> ♫Welcome to the jungle, it gets worse here every day♫
> ![image](https://user-images.githubusercontent.com/19940508/62349590-50439f00-b554-11e9-84cb-95c74e649526.png)

Hi, can you tell us more about the specifics?  Are these services or one-off tasks?  What placement strategy are you using?
Thanks

**Comment by benyanke** [comment]
Is there an ETA on re-balancing? I'm currently rolling out some clusters on mainly (or even exclusively) spot instances and I'd like to be able to easily rebalance workloads.


Similarly, it would be very useful to have a way to select containers for rebalancing - for example, a stateless middle tier app container or cron worker is far more adventageous for rescheduling than a relational database, which may incur downtime. If it was possible to have containers marked somehow (docker labels?) that they are safe to kill at will, provided the service stays in health limits, that would be ideal.

**Comment by hehnope** [comment]
tbh, we know the ETA is "when EKS is ready"

**Comment by benyanke** [comment]
so ECS is going out of support soon? Should I be stopping my investments into ECS?

**Comment by hehnope** [comment]
I doubt it's actually going out of support. Amazon is focused on EKS. I cannot see them really caring about ECS in the long-term.

**Comment by coultn** [comment]
ECS and EKS are both important to our customers and therefore both are important to AWS.  As you can see on the container roadmap, we have been continuously shipping features and improvements for both products.  As an example of the ongoing level of support, ECS and EKS also have similar runway on this roadmap. ECS has 4 coming soon, 11 we're working on it issues here.  Excluding region launches, EKS has 4 coming soon, 13 we're working on it. We listen closely to our customers, and this roadmap is one of the key ways we do that, so if you have feedback (such as have been given on the rebalancing issue) we want to hear it!

**Comment by benyanke** [comment]
Thanks @coultn !

**Comment by asdf01** [comment]
Hi @coultn Thanks for your interest in this issue.  

Firstly I hope you can understand my position if I come across a little pessimistic.  I've been around for a while and the only way to use AWS is to take what AWS gives you.  Asking and begging AWS for stuff is just a waste of time and setting yourself up for false hope.  These views were calcified over the years from painful episodes with AWS Support.  

Explaining our situation yet again to AWS would just create more false hope.  Most likely this exercise is just another complete waste of time, but hopefully someone would get some entertainment value out of it.  Please forgive the snarky comments... it's good for my mental health.

> can you tell us more about the specifics? 

~350 ecs services -> ~400 -> ~500 ecs tasks through the course of the day with autoscaling.
Is that all AWS was missing?  Specifics?  I think all of us in this thread is feeling a little silly right now.  If only one of us over the course of the last 4 years was a little more specific.  We could have avoided all this grief.

> Are these services or one-off tasks? 

They are all tasks generated from services.  Some configured with fixed numbers of tasks.  Some configured with autoscaling from 2-30 tasks per service depending on load.

> What placement strategy are you using?

I'm using the AZ Spread my buttER placement strategy.  You didn't think the cause of everyone's grief was that simple did you?  Not one person in the last 4 years thought of using the AZ spread placement strategy.  

Within a particular service, the AZ spread strategy works fine.  The joke's on you if that's what you based your feasibility study on.  Because when you go to production with your full suite of services, ECS just keeps on allocating tasks to the same machine regardless of how busy the machine is.  

Even a simple round robin task allocation method would have produced acceptable results.  Instead, you get this retarded black box ECS allocation algorithm.  AWS support seems to love using the terms, "not necessarily" and "not guaranteed" to describe the allocation behaviour when the more appropriate word is "retarded".  

Even massaging the task placement by hand doesn't work.  Manually killing a task on a very busy machine, just sees the new instance of that task started back up on the same already busy machine.  

> @coultn: ECS has 4 coming soon, 11 we're working on it issues here. 
> @coultn: EKS has 4 coming soon, 13 we're working on it. 
> @benyanke: Thanks @coultn !

How can you guys even understand each other?!  You guys both look caucasian, is there an English v2.0 out that I don't know about?  Where's the English v2.0 roadmap?  Why did you guys drop backward compatibility with English v1.0?

> so if you have feedback (such as have been given on the rebalancing issue) we want to hear it!

What are you guys going to do with the last 4 years of feedback?  Have a laugh over a bottle of wine?  You guys could have taken any one of the following interpretation of rebalancing:
* number of tasks
* reserved cpu
* reserved memory
* actual cpu 
* actual memory

Instead you guys chose to go with "nothing".  So I'd be interested in hearing what you guys plan to do with this round of feedback... another bottle of wine?

**Comment by hehnope** [comment]
While I don't want to make this a dumping ground; I can only agree with @asdf01. For example an issue I ran into last year: https://github.com/aws/amazon-ecs-init/issues/163, you guys released a security version which so happened to include a new feature; this of course brought down production and forced me to blacklist updating packages from you guys. Why you guys release minor versions that are generally _supposed_ to be bug fixes/security fixes as major features is beyond me.

This has been an issue for 4 years now; all I can do is post a relevant video: https://www.youtube.com/watch?v=5P2jJdrz9bY

Can we stop talking about it and just get the feature in? kthx

**Comment by ChintanMangukiya-OUP2** [comment]
Wow. Lots of discussion going on. I will try put my case in lesser words which should cover the pain area we all are facing with a solid example. 

I have 6 different types of services running total 10 tasks. 
2(7000MB each), 1(2048MB each), 2(768MB each), 2(768MB each), 2(128MB each), 1(512MB each). 
CPU reservation is not set. Not needed for me.  
m5.large(8GB and 2vCPU) instance used.
Desired cluster instances count is 3. Max 20. Using SchedulableContainers(Google it!) mechanism to auto scale instances. 
Default placement strategy with Roll Over.  

Results : 
**6 instances occupied, Worst case scenario :** with app changes deploying in random fashion. Normal case. Developer may commit change any time to any task. This remain for long time if no changes deployed or before ECS realizes to optimize it. Costs me more. 
**5 instances occupied, Average case scenario :** with app changes deploying in random fashion. Normal case. Developer may commit change any time to any task. This is general case most of the times. 
**4 instances occupied, Best case scenario :** If I manually launch above services starting from biggest to smallest taking good enough small pause. 

If someone can suggest a solution or AWS is happy to fix this problem then it is great. I don't think it is a big challenge for AWS. I don't mind if AWS will spin one more instance for some time to re-balance this. :)

Thanks for hearing me out. Cheers.

**Comment by cpmangukiya** [comment]
Wow. Lots of discussion going on. I will try put my case in lesser words which should cover the pain area we all are facing with a solid example. 

I have 6 different types of services running total 10 tasks. 
2(7000MB each), 1(2048MB each), 2(768MB each), 2(768MB each), 2(128MB each), 1(512MB each). 
CPU reservation is not set. Not needed for me.  
m5.large(8GB and 2vCPU) instance used.
Desired cluster instances is 3. Max 20(Just in case!). Using SchedulableContainers(Google it!) mechanism to auto scale instances. 
Default Placement Strategy with Roll Over.  

**Results :** 
**6 instances occupied, Worst case scenario :** with app changes deploying in random fashion. Normal case. Developer may commit change any time to any task. This remain for long time if no changes deployed or before ECS realizes to optimize it. Costs me more. 
**5 instances occupied, Average case scenario :** with app changes deploying in random fashion. Normal case. Developer may commit change any time to any task. This is general case most of the times. 
**4 instances occupied, Best case scenario :** If I manually launch above services starting from biggest to smallest taking good enough small pause. This remain for a while only until something is committed. 

If someone can suggest a solution or AWS is happy to fix this problem then it is great. I don't think it is a big challenge for AWS. I don't mind if AWS will spin one more instance for some time to re-balance this. :)

Thanks for hearing me out. Cheers.

**Comment by pavneeta** [comment]
https://github.com/aws/containers-roadmap/issues/136#issuecomment-568299433

**Comment by lofrank** [comment]
We still have this issue as of 2020.

![image](https://user-images.githubusercontent.com/3665054/74098698-96e20980-4b34-11ea-9bfd-41f43cd11226.png)

**Comment by heijmerikx** [comment]
![balancing](https://user-images.githubusercontent.com/4012/76309868-f8310000-62cd-11ea-81d3-045475d309a6.png)
I'm even considering one machine per cluster now...

**Comment by nathanpeck** [comment]
One solution that you can use today is the "cluster capacity providers" feature of ECS: run capacity providers with its own ASG for each availability zone that you want to run tasks in and then set even weights so that tasks get distributed evenly across all the capacity providers. This will ensure even distribution across all three availability zones at all times: https://docs.aws.amazon.com/AmazonECS/latest/developerguide/cluster-capacity-providers.html

For distribution across instances within a single ASG note that the default placement strategy is "random". Ensure that you are using "spread on instanceId" (https://docs.aws.amazon.com/AmazonECS/latest/developerguide/task-placement-strategies.html)  Note that each service gets spread individually though, so if you are running many services with only 1 task they will all get "spread" individually across the available pool of instances and their first task will likely be spread to the same first instance. This may explain some of the results some people are seeing with all tasks on a single instance.

**Comment by hlarsen** [comment]
There is a new feature in ECS preview to update placement strategies and constraints. The announcement says the changes will be enforced at the next deploy. Does that mean we can use this new feature to re-balance tasks that have been placed on sub-optimal hosts?

While not exactly the solution this issue is probably looking for, it sounds like this would allow one way of redistributing tasks for those who need it.

**Comment by sunomie** [comment]
Is there going to be any action on this or do we need to switch to Kubernetes?  This is a pretty major problem for a production environment.  Cluster capacity providers are a really big improvement, but it's not really a full solve for this.  As nathanpeck pointed out the tasks still pile up on certain instances.  If the answer is "we don't have resources to work on this and it's not going to get fixed anytime soon", that's fine and I get it; but please just say that so that we know that we need to make the investment to switch everything over to Kubernetes.

**Comment by pavneeta** [comment]
> There is a new feature in ECS preview to update placement strategies and constraints. The announcement says the changes will be enforced at the next deploy. Does that mean we can use this new feature to re-balance tasks that have been placed on sub-optimal hosts?
> 
> While not exactly the solution this issue is probably looking for, it sounds like this would allow one way of redistributing tasks for those who need it.

Hi @hlarsen Yes, you can use the Update Placement strategy feature to re-balance a given service acorss ECS instances that have piled up on certain instances. You can use `--Force-Deploy` flag to cause an immediate deployment and enforcement of the updated palcement configuration or wait for the next service deployment. 

@sunomie - I understand your frustration, unfortunately there is no one size fits all solution here. Could you please share more information about your cluster setup or discuss the problem in depth offline, you can reach me at pavneeta@amazon.com
- Are there multiple services deployed on it ?
-  Are you using Capacity Providers with Cluster Auto Scaling ?
- Were you able to use the update placement strategy to redistribute tasks ? 
- Is the goal to improve cluster utilization or heat management ?

**Comment by nam178** [comment]
I cannot believe this issue has not been resolved after 5 years. Where is your customer obsession now?

I'm emailing Bezos.

**Comment by jamengual** [comment]
> cluster capacity providers

yep, you could use that but you need a math degree to understand what is going to do and even then the provider examples do not work

**Comment by Makeshift** [comment]
Happy fifth birthday, issue #105! 🥳  I've been subscribed to this thread for four of those years now!

**Comment by chrisj-au** [comment]
5 years aws, really?

**Comment by brendan-sherrin** [comment]
How is this still a thing after 5 years...

**Comment by LordSnooz** [comment]
+1 

We sell the autoscaling of EC2 instances, but the solution to rebalance the tasks still does not exist ... Ultimately, autoscaling is useless in certain use cases! Need rebalancing ASAP on autoscaling up!

**Comment by acashley** [comment]
+1

**Comment by asdf01** [comment]
Jeff Bezos is in space already.  Fix it.

**Comment by m-b-rashid** [comment]
+1

**Comment by cpmangukiya** [comment]
Hey AWS, come on, consider the proposal. 

We mostly want that AWS automatically (with a single flag set!) rebalance tasks such that they are optimally distributed on minimum number of instances for **a longer duration**. AWS may occasionally spin up/shut down extra instances (in-between) for a short time just to rebalance tasks and we would be ok to pay few extra minutes of instances. 

May be say Yes or No or something else. ;)

**Comment by dkallakuri** [comment]
Thanks for the feedback. We are looking into this and have it on our roadmap. In interim, we will look into providing a workaround.

**Comment by psyhomb** [comment]
> Thanks for the feedback. We are looking into this and have it on our roadmap. In interim, we will look into providing a workaround.

~5 years ago https://github.com/aws/containers-roadmap/issues/105#issuecomment-148430229

**Comment by psyhomb** [comment]
Also in order to mitigate balancing problem with the cluster you should also consider implementing a simple sort algorithm per available resources (cluster constraints - satisfy CPU, memory and port requirements), that way a new task will always be started on the instance that has most available resources and not on the instance that is just suitable.

**Comment by acashley** [comment]
New year, old problem. C'mon AWS ECS team...

**Comment by lobsterdore** [comment]
This issue should be handled natively, but I thought I'd add in how I solved this at the place I work at.

I put together a Lambda that runs every few minutes, it picks out the least used container instance in a cluster and checks to see if its tasks can fit onto the other instances, if they can fit it simply sets the instance state to draining. Once all tasks are drained auto-scaling eventually terminates the instance.

This is a fairly straightforward that does not involve manipulating any of the services themselves.

**Comment by totojack** [comment]
+1 for rebalancing

**Comment by wesctl** [comment]
The easiest way to solve this issue and cost optimize via bin packing is to utilize Spot.io [Ocean for ECS ](https://docs.spot.io/ocean/overview-ecs)--> https://docs.spot.io/ocean/overview-ecs
Hit us up if you want a demo!

**Comment by GreasyAvocado** [comment]
Any update on this?

**Comment by dinaant** [comment]
+1 for rebalancing

**Comment by jdaviderb** [comment]
+1

**Comment by taras-mrtn** [comment]
+1 for rebalancing

**Comment by jakeinflab** [comment]
You can solve the container imbalance issue caused by rolling updates by creating an Auto Scaling Group for each Availability Zone (AZ) and dividing the Capacity Provider for each AZ. This can resolve the immediate problem, but it would be nice if a better method could be suggested, as managing this setup can be difficult. :)
https://containersonaws.com/pattern/balanced-ecs-on-ec2-container-deployment-across-az

**Comment by Nevon** [comment]
It's definitely a good feature and solves one of the issues brought up in this thread. I haven't had a chance to test it out yet, but I will as soon as possible. However, it doesn't fully solve the general problem as it only addresses the specific placement strategy of AZ spread. Other strategies, such as the one brought up in the issue description that talks about inefficient resource utilization when using a binpacking strategy, are not resolved by this improvement.

**Comment by AbhishekNautiyal** [comment]
This capability has been delivered with the Amazon ECS Managed Instances launch, as covered in https://github.com/aws/containers-roadmap/issues/2452#issuecomment-3353559697

You can read about task placement and cost optimization in our [documentation](https://docs.aws.amazon.com/AmazonECS/latest/developerguide/managed-instances-infrastructure-optimization.html).  Please also look out for a blog post coming soon for a deeper dive on autoscaling and optimization with ECS Managed Instances.
